### PR TITLE
fix: checksums are now fixed for similar ledgerIDs, ex: 0x01 and 0x10

### PR DESCRIPTION
**Pull request by andrix10**
Signed-off-by: Andrei Kuzmiankov <andrei@launchbadge.com>

closes: #417

### CODE REVIEW

File: entity_id.go [hunk 1]
```diff
@@ -94,17 +100,14 @@ func _CheckChecksum(ledgerID string, address string) string {
 	asciiA := []rune("a")[0]
 	w := 31
 
-	id := ledgerID + "000000000000"
+	zeros, _ := hex.DecodeString("000000000000")
+
+	id := append(ledgerID, zeros...)
```
Comment (danielakhterov) [inline entity_id.go:105]:
```suggestion
	id := append(ledgerID, []byte{0, 0, 0, 0, 0, 0})
```

File: account_id_unit_test.go [hunk 2]
```diff
@@ -35,3 +35,12 @@ func TestUnitAccountIDFromStringAlias(t *testing.T) {
...
 	assert.Equal(t, id.String(), id2.String())
 }
+
+func TestUnitChecksum(t *testing.T) {
+	ad1, err := _ChecksumParseAddress("1", "0.0.3")
+	require.NoError(t, err)
+	ad2, err := _ChecksumParseAddress("10", "0.0.3")
+	require.NoError(t, err)
+
+	require.NotEqual(t, ad1.correctChecksum, ad2.correctChecksum)
```
Comment (danielakhterov) [inline account_id_unit_test.go:45]: Is this actually the case? I don't think we want 1 and 10 to equal the same checksum anymore, we actually want it to be different now.

**Review by danielakhterov — CHANGES_REQUESTED** [review]

File: entity_id.go [hunk 3]
```diff
@@ -94,17 +100,15 @@ func _CheckChecksum(ledgerID string, address string) string {
 	asciiA := []rune("a")[0]
 	w := 31
 
-	id := ledgerID + "000000000000"
+	zeros, _ := hex.DecodeString("000000000000")
```
Comment (danielakhterov) [inline entity_id.go:103]: There is no point in decoding a string of 0's. If you know the value you can just hard code it. In fact you might be able to create a `const` for this.

**Review by danielakhterov — CHANGES_REQUESTED** [review]

File: entity_id.go [hunk 4]
```diff
@@ -94,17 +100,14 @@ func _CheckChecksum(ledgerID string, address string) string {
...
+
+	id := append(ledgerID, zeros...)
 	h := make([]int64, 0)
 
-	for i := 0; i < len(id); i += 2 {
-		processed, _ := strconv.ParseInt(id[i:i+2], 16, 64)
+	for i := 0; i < len(id); i += 1 {
+		processed, _ := strconv.ParseInt(hex.EncodeToString(id[i:i+1]), 16, 64)
```
[the remark this answers is not in the document]
Comment (andrix10) [reply entity_id.go:110]: I am confused what yoiu mean, im converting it from base 16, and bit size 64 to int64, so far this is the only way I know how to do that.

File: account_id_unit_test.go (hunk 2 above)
Comment (andrix10) [reply account_id_unit_test.go:45]: Ye this is the test for that.

File: entity_id.go (hunk 3 above)
Comment (andrix10) [reply entity_id.go]: Wait i didn't see this comment yesterday some reason, but yes its fixed.

File: ledger_id.go [hunk 5]
```diff
@@ -26,7 +26,11 @@ func LedgerIDFromString(id string) (*LedgerID, error) {
 			_LedgerIDBytes: []byte{2},
 		}, nil
 	default:
-		temp, err := hex.DecodeString(id)
+		id2 := id
+		if len(id)%2 != 0 {
+			id2 = "0" + id
```
Comment (danielakhterov) [inline ledger_id.go:31]: Don't do this, let it die in the `hex.DecodeString()` instead.

**Review by danielakhterov — CHANGES_REQUESTED** [review]

**Pull request merged**